mentat/code_context.py

```diff
@@ -271,37 +273,132 @@ def _get_all_features(
-        # and can't return it as an invalid path
-        paths, invalid_paths = get_include_files([path], [])
-        removed_paths = list[Path]()
-        for new_path in paths.keys():
-            if new_path in self.include_files:
-                removed_paths.append(new_path)
-                del self.include_files[new_path]
-        return removed_paths, invalid_paths
+    def include(
+        self, path: Path | str, ignore_patterns: Iterable[Path | str] = []
+    ) -> Set[Path]:
+        """Add code to the context
+
+        '.' is replaced with '*' (recusively search the cwd)
+
+        Args:
+            `path`: can be a relative or absolute file path, file interval path, directory, or glob pattern.
+
+        Return:
+            A set of paths that have been successfully added to the context
+        """
+        session_context = SESSION_CONTEXT.get()
+
+        path = Path(path)
+        if str(path) == ".":
+            path = Path("*")
+
+        included_paths: Set[Path] = set()
+        try:
+            code_features = get_code_features_for_path(
+                path=path,
+                cwd=session_context.cwd,
+                ignore_patterns=[
+                    *ignore_patterns,
+                    *self.ignore_patterns,
+                    *session_context.config.file_exclude_glob_list,
+                ],
+            )
+        except PathValidationError as e:
+            session_context.stream.send(e, color="light_red")
+            return included_paths
+
+        for code_feature in code_features:
+            if code_feature.path in self.include_files:
+                self.include_files[code_feature.path].append(code_feature)
+            else:
+                self.include_files[code_feature.path] = [code_feature]
+
+        return included_paths
+
+    def exclude(self, path: Path | str) -> Set[Path]:
+        """Remove code from the context
+
+        # NOTE: should this be the intended behavior?
+        Code Features are removed from the context and are excluded from any future `include` calls (unless explicity specified).
+        For paths that are files or file intervals, if they don't already exist in the Code Context an Exception is thrown.
+        For file intervals, if the interval doesn't *exactly match* any intervals in the code context, nothing is removed.
+
+        Args:
+            `path`: can be a relative or absolute file path, file interval path, directory, or glob pattern.
+
+        Return:
+            A set of paths that have been successfully removed from the context
+        """
+        session_context = SESSION_CONTEXT.get()
+
+        path = Path(path)
+
+        excluded_paths: Set[Path] = set()
+
+        try:
+            validated_path = validate_and_format_path(
+                path, session_context.cwd, check_for_text=False
+            )
+            # file
+            if validated_path.is_file():
```

Comment:
We use this exact set of file, file interval, directory, glob pattern if statements 3 or 4 times; there shouldn't need to be a difference at all. Why don't we do it like we did before where we just call get_include_files on the path and include_files handled everything in one function?

Reply:
>there shouldn't need to be a difference at all

Could you elaborate on this? Not sure I understand